tools/ReadConsoleInputStream/ReadConsoleInputStream.cs

```diff
@@ -0,0 +1,194 @@
+        private IntPtr _handle;
+
+        /// <summary>
+        /// Creates an instance of ReadConsoleInputStream over the standard handle for StdIn. 
+        /// </summary>
+        /// <param name="nonKeyEvents">A BlockingCollection provider/consumer collection for collecting non key events.</param>
+        public ReadConsoleInputStream(BlockingCollection<Kernel32.INPUT_RECORD> nonKeyEvents) :
+            this(Kernel32.GetStdHandle(Kernel32.StdHandleType.STD_INPUT_HANDLE), nonKeyEvents)
+        {
+        }
+
+        /// <summary>
+        /// Creates an instance of ReadConsoleInputStream over a caller-provided standard handle for stdin. 
+        /// </summary>
+        /// <param name="handle">A HFILE handle representing StdIn</param>
+        /// <param name="nonKeyEvents">A BlockingCollection provider/consumer collection for collecting non key events.</param>
+        internal ReadConsoleInputStream(HFILE handle,
+            BlockingCollection<Kernel32.INPUT_RECORD> nonKeyEvents)
+        {
+            Debug.Assert(handle.IsInvalid == false, "handle.IsInvalid == false");
+
+            _handle = handle.DangerousGetHandle();
+            _nonKeyEvents = nonKeyEvents;
+        }
+
+        public override bool CanRead { get; } = true;
+
+        public override bool CanWrite => false;
+
+        public override bool CanSeek => false;
+
+        public override long Length => throw new NotSupportedException("Seek not supported.");
+
+        public override long Position
+        {
+            get => throw new NotSupportedException("Seek not supported.");
+            set => throw new NotSupportedException("Seek not supported.");
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            _handle = IntPtr.Zero;
+
+            base.Dispose(disposing);
+        }
+
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+            ValidateRead(buffer, offset, count);
+
+            Debug.Assert(offset >= 0, "offset >= 0");
+            Debug.Assert(count >= 0, "count >= 0");
+            Debug.Assert(buffer != null, "bytes != null");
+
+            // Don't corrupt memory when multiple threads are erroneously writing
+            // to this stream simultaneously.
+            if (buffer.Length - offset < count)
+                throw new IndexOutOfRangeException("IndexOutOfRange_IORaceCondition");
+
+            int bytesRead;
+            int ret;
+
+            if (buffer.Length == 0)
+            {
+                bytesRead = 0;
+                ret = Win32Error.ERROR_SUCCESS;
+            }
+            else
+            {
+                var charsRead = 0;
+                bytesRead = 0;
+
+                var records = new Kernel32.INPUT_RECORD[BufferSize];
+
+                // begin input loop
+                waitForInput:
+
+                var readSuccess = Kernel32.ReadConsoleInput(_handle, records, 256, out var recordsRead);
```

Comment:
>256 [](start = 78, length = 3)

here's the magic 256 again. Can this be looked up based on the array size at runtime instead of magic constant?

Reply:
Good catch.